packages/react-dom-bindings/src/server/fizz-instruction-set/ReactDOMFizzInstructionSet.js

```diff
@@ -209,6 +217,50 @@ export function completeBoundary(suspenseBoundaryID, contentID, errorDigest) {
   }
 }
 
+export function completeContainer(
+  containerID,
+  contentID,
+  optionalErrorMessage,
+) {
+  const contentNode = document.getElementById(contentID);
+  // We'll detach the content node so that regardless of what happens next we don't leave in the tree.
+  // This might also help by not causing recalcing each time we move a child from here to the target.
+  contentNode.parentNode.removeChild(contentNode);
+
+  // Find the container node.
+  const containerNode = document.getElementById(containerID);
+  if (!containerNode) {
+    // The user must have already navigated away from this tree.
+    // E.g. because the parent was hydrated. That's fine there's nothing to do
+    // but we have to make sure that we already deleted the container node.
+    return;
+  }
+  // This container may have a stashed promise from the style insertions.
+  // We assume this and clear it now that the root is unblocked and hydration
+  // can continue. If there is an error message put it in this slot. The client
+  // will fall back to client rendering the root and use this message in the
+  // logging
+  if (optionalErrorMessage) {
+    throw optionalErrorMessage;
+  }
+
+  // If the container node has already been passed to createRoot we assume
+  // the client will render the root and we never swap the content in.
+  // Additionally if there was an error in loading styles we leave the fallback
+  // content in place and expect the client will fall back to client rendering
+  if (containerNode['_c']) {
+    return;
+  }
```

Comment:
This is not actually implemented yet. It may not be necessary.

1. we won't support calling `createRoot` when you are also streaming with `renderIntoContainer` so there should be no time when the root renders before `completeContainer` is called
2. for `hydrateRoot` the root suspends with delay but because no other updates can be scheduled until the root stops suspending it should delay until this function returns (resolved) or throws (rejected)
3. If you use createRoot in fallback bootstrap scripts we are never going to reveal content anyway b/c the style insertion or completeContainer never get called

If the above expressions are correct (I think they are but would like help reasoning about this) then we don't need a back-signal on the container that causes a content reveal bailout

Reply:
outdated